File: ml/process_data.py
from __future__ import annotations
import json
import os
from dataclasses import dataclass, field
from typing import ClassVar

DATA_DIRECTORY = "../data/unprocessed/"
MAX_CONVERSATION_LENGTH = 20
ONE_HOUR = 3600000

@dataclass
class Message:
    role: str
    context: str
    content: str

@dataclass
class MessageContext:
    replied_user: str | None
    replied_message_id: int | None
    id: int = -1

    def format(self) -> str:
        string = f"Message id {self.id}."
        if self.replied_user is not None:
            if self.replied_message_id is not None:
                string += f" Replying to message id {self.replied_message_id} by {self.replied_user}."
            else:
                string += f" Replying to an unknown message by {self.replied_user}."

        return string

@dataclass
class RawMessage:
    author_id: str
    content: str
    message_id: str
    timestamp: int
    replied_message_id: str | None
    replied_user_id: str | None

    author: str | None = None
    context: MessageContext = field(default_factory=lambda: MessageContext(None, None))

    def format(self) -> str:
        return Message(self.author if self.author is not None else self.author_id, self.context.format(), self.content).__str__()
    
@dataclass
class Prompt:
    instruction: str
    messages: list[RawMessage]
    
    _next_user_number: int = 1
    _user_id_map: dict[str, str] = field(default_factory=dict)

    _next_message_id: int = 1
    _message_id_map: dict[str, int] = field(default_factory=dict)

    def __post_init__(self):
        for message in self.messages:
            if message.author is not None:
                print("error: iterating over messages that have already been processed")
                continue

            # Set author username
            if message.author_id in self._user_id_map:
                message.author = self._user_id_map[message.author_id]
            else:
                message.author = f"User {self._next_user_number}"
                self._next_message_id += 1
                self._user_id_map[message.author_id] = message.author

            # Set message context
            message.context.id = self._next_message_id
            self._next_message_id += 1
            self._message_id_map[message.message_id] = message.context.id

            # Set replied message ids in context
            if message.replied_message_id is not None:
                if message.replied_message_id in self._message_id_map:
                    message.context.replied_message_id = self._message_id_map[message.replied_message_id]

                if message.replied_user_id in self._user_id_map:
                    message.context.replied_user = self._user_id_map[message.replied_user_id]

    def format(self):
        formatted_messages: list[str] = []
        for raw_message in self.messages:
            formatted_messages.append(raw_message.format())

        return {
            "instruction": self.instruction,
            "messages": formatted_messages
        }



def load_message_data(directory_path) -> list[dict[str, str]]:
    messages: list = []

    if not os.path.exists(directory_path):
        print(f"Directory \'{directory_path}\' not found.")
        return messages
    
    for filename in os.listdir(directory_path):
        if not filename.endswith(".json"):
            continue
        file_path = os.path.join(directory_path, filename)
        try:
            with open(file_path, "r", encoding="utf-8") as file:
                data: dict = json.load(file)
                if isinstance(data, list):
                    messages.extend(data)
                    print(f"Loaded {filename} with {len(data)} messages")
        except Exception as e:
            print(f"Error loading {filename}, {type(e).__name__} - {e}")

    return messages

def process_messages_into_conversations(data: list) -> list[Prompt]:
    prompts: list[Prompt] = []
    raw_messages: list[RawMessage] = []
    start_index: int = 0

    def process_conversation_segment(start_index: int, end_index: int):
        conversation_length = end_index - start_index
        
        if conversation_length <= 0:
            return

        # Full conversation prompt
        if conversation_length != 1:
            formatted_full_messages = [msg for msg in raw_messages[start_index:end_index]]
            user = raw_messages[end_index - 1].author
            
            instruction = f"You are {user} engaging in a conversation on Discord."
            prompts.append(Prompt(instruction, formatted_full_messages))
        
        # First message only (conversation starter)
        first_message = [raw_messages[start_index]]
        first_instruction = f"You are {raw_messages[start_index].author} starting a conversation on Discord."
        prompts.append(Prompt(first_instruction, first_message))

        # Half conversation
        if conversation_length > 10:
            half_length = conversation_length // 2
            half_end_index = start_index + half_length
            formatted_half_messages = [msg for msg in raw_messages[start_index:half_end_index]]
            half_user = raw_messages[half_end_index - 1].author
            half_instruction = f"You are {half_user} engaging in a conversation on Discord."
            prompts.append(Prompt(half_instruction, formatted_half_messages))

    total_length = len(data)

    for i, message in enumerate(data):
        raw_message = dict_to_raw_message(message)
        
        if str(raw_message.content).strip() == "":
            continue
            
        raw_messages.append(raw_message)
        
        if len(raw_messages) > 1 and (
            raw_message.timestamp - raw_messages[-2].timestamp >= ONE_HOUR or 
            len(raw_messages) - start_index > MAX_CONVERSATION_LENGTH
        ):

            process_conversation_segment(start_index, len(raw_messages) - 1)
            
            start_index = len(raw_messages) - 1
        
        # Debug
        print(f"Processed {i+1} messages out of {total_length}")
    
    # Process the final conversation
    if len(raw_messages) > start_index:
        process_conversation_segment(start_index, len(raw_messages))
        
    return prompts

#TODO: fix message and user ids - this is from the saving stuff
def save_prompts_to_jsonl(prompts: list[Prompt]) -> None:
    output_path = "../data/processed/prompts.jsonl"

    os.makedirs(os.path.dirname(output_path), exist_ok=True)

    with open(output_path, "w") as file:
        for prompt in prompts:
            prompt_string = prompt.format()
            file.write(json.dumps(prompt_string) + "\n")
            pass

    print(f"Saved {len(prompts)} prompts to {output_path}")

def dict_to_raw_message(message: dict) -> RawMessage:
    required_fields = ["authorId", "content", "id", "createdTimestamp"]
    for field in required_fields:
        if field not in message:
            raise ValueError(f"Missing required field: {field}")
    
    replied_message_id = None
    if message.get("reference") is not None:
        replied_message_id = message["reference"].get("messageId")
    
    replied_user_id = None
    if message.get("mentions") is not None:
        replied_user_id = message["mentions"].get("repliedUser")
    
    return RawMessage(
        message["authorId"],
        message["content"], 
        message["id"], 
        message["createdTimestamp"],
        replied_message_id,
        replied_user_id
    )

def main():
    messages = load_message_data(DATA_DIRECTORY)
    if not messages:
        print("No messages found to process.")
        return
        
    prompts = process_messages_into_conversations(messages)
    save_prompts_to_jsonl(prompts)
    print(f"Processing complete: {len(prompts)} prompts generated.")

if __name__ == "__main__":
    main()

File: ml/process_data.ts
import * as fs from 'fs';
import * as path from 'path';

const DATA_DIRECTORY = "./data/unprocessed/";
const MAX_CONVERSATION_LENGTH = 20;
const ONE_HOUR = 3600000;

class Message {
    private role: string;
    private context: string;
    private content: string;

    public constructor(role: string, context: string, content: string) {
        this.role = role;
        this.context = context;
        this.content = content;
    }

    public toJSON() {
        return {
            role: this.role,
            context: this.context,
            content: this.content
        };
    }
}

class User {
    private id: string;
    private name: string | undefined;

    public constructor(id: string) {
        this.id = id;
    }

    setName(name: string): void {
        if (this.name) return;
        this.name = name;
    }

    getName(): string | undefined {
        return this.name;
    }
}

class RawMessage {
    private readonly authorId: string
    private readonly content: string
    private readonly discordMessageId: string
    private readonly timeStamp: number
    private repliedMessageId?: string
    private repliedUserId?: string
    private context: MessageContext | undefined

    public constructor(authorId: string, content: string, discordMessageId: string, timeStamp: number, repliedUserId?: string, repliedMessageId?: string) {
        this.authorId = authorId;
        this.content = content;
        this.discordMessageId = discordMessageId;
        this.timeStamp = timeStamp;
        this.repliedUserId = repliedUserId;
        this.repliedMessageId = repliedMessageId;
    }

    public getAuthorId(): string {
        return this.authorId;
    }

    public getContent(): string {
        return this.content;
    }

    public getDiscordMessageId(): string {
        return this.discordMessageId;
    }

    public getTimeStamp(): number {
        return this.timeStamp;
    }

    public getRepliedMessageId(): string | undefined {
        return this.repliedMessageId;
    }

    public setRepliedMessageId(repliedMessageId: string): void {
        this.repliedMessageId = repliedMessageId;
    }

    public getRepliedUserId(): string | undefined {
        return this.repliedUserId;
    }

    public setRepliedUserId(repliedUserId: string): void {
        this.repliedUserId = repliedUserId;
    }

    public getContext(): MessageContext | undefined {
        return this.context;
    }

    public setContext(context: MessageContext): void {
        // Defensive copy - not an inefficiency
        this.context = new MessageContext(
            context.getId(),
            context.getRepliedMessageId(),
            context.getRepliedUser()
        );
    }
}

class MessageContext {
    private readonly id: number;
    private readonly repliedMessageId?: number;
    private readonly repliedUserId?: number;

    public constructor(id: number, repliedMessageId?: number, repliedUserId?: number) {
        this.id = id;
        this.repliedMessageId = repliedMessageId;
        this.repliedUserId = repliedUserId;
    }

    public toString(): string {
        let string: string = "";
        string += `Message ID ${this.id}.`;

        if (this.repliedMessageId || this.repliedUserId) {
            string += `\nReplying to`;

            if (this.repliedMessageId) {
                string += ` message ID ${this.repliedMessageId}`;
            } else {
                string += " an unknown message";
            }

            string += " by";

            if (this.repliedUserId) {
                string += ` ${toUsername(this.repliedUserId)}`;
            } else {
                string += " an unknown user";
            }
        }

        return string;
    }

    public getId(): number {
        return this.id;
    }

    public getRepliedMessageId(): number | undefined {
        return this.repliedMessageId;
    }

    public getRepliedUser(): number | undefined {
        return this.repliedUserId;
    }
}

class Prompt {
    private instruction: string;
    private readonly messages: ReadonlyArray<RawMessage>;

    private nextUserNumber: number = 1;
    private userIdMap: Map<string, number> = new Map<string, number>();

    private nextMessageNumber: number = 1;
    private messageIdMap: Map<string, number> = new Map<string, number>();

    public constructor(messages: Array<RawMessage>) {
        this.messages = [...messages];

        let userId: number;
        this.messages.forEach((message) => {
            if (this.userIdMap.has(message.getAuthorId())) {
                userId = this.userIdMap.get(message.getAuthorId())!;
            } else {
                userId = this.newUserId();
                this.userIdMap.set(message.getAuthorId(), userId);
            }

            if (!message.getContext()) {
                this.initialiseMessageContext(message);
            }
        });

        const lastMessageUsername = toUsername(this.userIdMap.get(messages.at(-1)!.getAuthorId())!);
        let instructionBuilder = `You are ${lastMessageUsername} `;
        if (messages.length === 1) {
            instructionBuilder += "starting";
        } else {
            instructionBuilder += "engaging in";
        }
        instructionBuilder += " a conversation on Discord";

        this.instruction = instructionBuilder;
    }

    public toString(): string {
        const historyRawMessages = this.messages.slice(0, -1);
        const historyMessages = new Array<Message>();
        historyRawMessages.forEach((message) => {
            let formattedContext: string;
            let formattedUsername: string;

            if (message.getContext()) {
                formattedContext = message.getContext()!.toString();
            } else {
                throw new Error(`Context for message undefined!`);
            }

            if (this.userIdMap.has(message.getAuthorId())) {
                formattedUsername = toUsername(this.userIdMap.get(message.getAuthorId())!);
            } else {
                throw new Error(`User ID ${message.getAuthorId()} not found in userIdMap`);
            }

            historyMessages.push(new Message(formattedUsername, formattedContext, message.getContent()));
        });

        const object = {
            instruction: this.instruction,
            context: historyMessages,
            response: this.messages.at(-1)!.getContent(),
            responseUser: this.userIdMap.get(this.messages.at(-1)!.getAuthorId())
        };

        return JSON.stringify(object);
    }

    private initialiseMessageContext(message: RawMessage) {
        const newId = this.newMessageId();

        this.messageIdMap.set(message.getDiscordMessageId(), newId);
        
        if (message.getRepliedUserId() && message.getRepliedMessageId()) {
            if (this.userIdMap.has(message.getRepliedUserId()!) && this.messageIdMap.has(message.getRepliedMessageId()!)) {
                message.setContext(new MessageContext(
                    newId,
                    this.messageIdMap.get(message.getRepliedMessageId()!)!,
                    this.userIdMap.get(message.getRepliedUserId()!)!
                ));
            } else if (this.userIdMap.has(message.getRepliedUserId()!)) {
                message.setContext(new MessageContext(
                    newId,
                    undefined,
                    this.userIdMap.get(message.getRepliedUserId()!)!
                ));
            } else {
                // console.warn("Reply message outside of scope found");
                message.setContext(new MessageContext(newId, undefined, undefined));
            }
        } else {
            message.setContext(new MessageContext(newId, undefined, undefined));
        }
    }

    private newMessageId(): number {
        return this.nextMessageNumber++;
    }

    private newUserId(): number {
        return this.nextUserNumber++;
    }
}

function toUsername(id: number): string {
    return `User ${id}`;
}

function loadMessageData(directoryPath: string): Array<RawMessage> {
    const messages: Array<RawMessage> = [];

    if (!fs.existsSync(directoryPath)) {
        console.log(`Directory '${directoryPath}' not found.`);
        return messages;
    }

    const files = fs.readdirSync(directoryPath);
    for (const filename of files) {
        if (!filename.endsWith('.json')) {
            continue;
        }

        const filePath = path.join(directoryPath, filename);

        try {
            const fileContent = fs.readFileSync(filePath, 'utf-8');
            const data = JSON.parse(fileContent);

            if (Array.isArray(data)) {
                const validMessages = data
                    .filter(msg => msg.content && !msg.system)
                    .map(msg => new RawMessage(
                        String(msg.authorId),
                        msg.content,
                        msg.id,
                        msg.createdTimestamp,
                        msg.mentions?.repliedUser || undefined,
                        msg.reference?.messageId || undefined
                    ));

                messages.push(...validMessages);
                console.log(`Loaded ${filename} with ${validMessages.length} valid messages`);
            }
        } catch (e) {
            console.log(`Error loading ${filename}, ${e instanceof Error ? e.name : 'Unknown'} - ${e instanceof Error ? e.message : e}`);
        }
    }

    return messages;
}

function generatePrompts(messages: RawMessage[]): Prompt[] {
    let startIndex = 0;
    let tempConversation: Array<RawMessage> = [];
    const prompts = new Array<Prompt>();

    for (let index = 0; index < messages.length; index++) {
        tempConversation.push(messages[index]);
        const conversationLength = index - startIndex;

        if (index === startIndex) {
            prompts.push(new Prompt(tempConversation));
            continue;
        }

        if (messages[index].getTimeStamp() - messages[index - 1].getTimeStamp() > ONE_HOUR) {
            addHalfPrompt(conversationLength);
            pushPrompt();
            continue;
        }

        if (conversationLength >= MAX_CONVERSATION_LENGTH) {
            addHalfPrompt(conversationLength);
            pushPrompt();
            continue;
        }

        function pushPrompt() {
            prompts.push(new Prompt(tempConversation));
            tempConversation = [];
            startIndex = index + 1;
        }
    }

    function addHalfPrompt(conversationLength: number) {
        if (conversationLength > 10) {
            const halfLength = Math.round(conversationLength / 2);
            const halfEndIndex = startIndex + halfLength;
            const halfMessages = messages.slice(startIndex, halfEndIndex);
            prompts.push(new Prompt(halfMessages));
        }
    }

    return prompts;
}

function savePromptsToJsonl(prompts: Prompt[]) {
    const outputDir = './data/processed';
    const outputFile = path.join(outputDir, 'prompts.jsonl');

    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
    }

    let outputContent = '';
    prompts.forEach(prompt => {
        outputContent += prompt.toString() + '\n';
    });

    fs.writeFileSync(outputFile, outputContent);
    console.log(`Saved ${prompts.length} prompts to ${outputFile}`);
}

function main() {
    savePromptsToJsonl(generatePrompts(loadMessageData(DATA_DIRECTORY)));
}

main();
